# Analysis Goal :

To perform hyper-parameter tuning using HyperOpt for SVM model

# Hyperparameter Tuning with hyperopt

Hyperparameter tuning is an important step for maximizing the performance of a model. Hyperparameters are certain values/weights that determine the learning process of an algorithm. Several Python packages have been developed specifically for this purpose. Scikit-learn provides a few options, GridSearchCV and RandomizedSearchCV being two of the more popular options. Outside of scikit-learn, the Optunity, Spearmint and hyperopt packages are all designed for optimization. In this task, we will focus on the hyperopt package which provides algorithms that are able to outperform randomized search.

In [1]:
from SvmModel import Model

In [2]:
model = Model("data_transformed.csv",
        "data_transformed.csv")

In [3]:
import numpy as np
import pandas as pd
from sklearn import svm
from sklearn.metrics import f1_score
from hyperopt import space_eval
from hyperopt import hp, fmin, tpe, Trials, STATUS_OK

In [4]:
import warnings
warnings.filterwarnings('ignore')

In [5]:
class controller():
    def __init__(self,model):
        self.model = model
    
    def optimize_hyperparam(self, n_eval=10):
        
        def objective(space4svm):
            f1 = model.svm(space4svm)
            print(f1)
            return {'loss': -f1, 'status': STATUS_OK}
        
        def hyper_params():
            space4svm ={
            'C': hp.uniform('C', 0, 20),
            'kernel': hp.choice('kernel', ['linear', 'sigmoid', 'poly', 'rbf']),
            'gamma': hp.choice('gamma', ['scale', 'auto']),
            'degree': hp.quniform('degree', 3, 10,1),
            'coef0': hp.uniform('coef0', 0,10)
            }
            return space4svm
        space4svm = hyper_params()
    
    
        # Initialize trials object.
        trials = Trials()
        #using seed to get repeatable results.
        seed = 123
        # run the hyper paramter tuning.
        best = fmin(fn=objective,   
                        space=space4svm,
                        algo=tpe.suggest,
                        max_evals=n_eval,
                        trials=trials,
                        rstate=np.random.RandomState(seed))
        print (space_eval(space4svm,best))
        hyperparams = space_eval(space4svm, best)
        return hyperparams 

In [6]:
hyperCl= controller(model)

In [7]:
hyperCl.optimize_hyperparam()

0.6984886458570669                                                                                                     
0.6984886458570669                                                                                                     
0.5989974937343359                                                                                                     
0.6992481203007519                                                                                                     
0.6992481203007519                                                                                                     
0.5088638883366223                                                                                                     
0.6992481203007519                                                                                                     
0.4029850746268657                                                                                                     
0.6992481203007519                      

{'C': 1.7718619582441852,
 'coef0': 1.6216340381955197,
 'degree': 8.0,
 'gamma': 'scale',
 'kernel': 'poly'}

# Conclusion:

This task is done in a team of 2 students. The given dataset was analyzed and modelled using SVM Model. Hyperparameters were tuned using hyperopt. Hyperparameter tuning is an important step in building a learning algorithm model. Best parameters for SVM model are 'C': 1.7718619582441852, 'coef0': 1.6216340381955197, 'degree': 8.0, 'gamma': 'scale', 'kernel': 'poly'